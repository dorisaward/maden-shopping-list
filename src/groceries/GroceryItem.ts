export type GroceryItem = {
  /**
   * Unique identifier for the grocery item
   */
  id: string;

  /**
   * The human readable name for the grocery item
   */
  name: string;

  /**
   * The nummber of the groceries to display on the list
   */
  quantity: number;
};
