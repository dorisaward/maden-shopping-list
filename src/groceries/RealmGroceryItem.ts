import Realm, {ObjectSchema} from 'realm';

export class RealmGroceryItem extends Realm.Object<RealmGroceryItem> {
  _id!: Realm.BSON.ObjectId;
  name!: string;
  quantity!: number;
  static schema: ObjectSchema = {
    name: 'RealmGroceryItem',
    properties: {
      _id: 'objectId',
      name: 'string',
      quantity: 'int',
    },
    primaryKey: '_id',
  };
}
