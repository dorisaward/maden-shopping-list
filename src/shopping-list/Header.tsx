import {Text, View} from 'react-native';
import {styles} from '../styles';
import React from 'react';

export const Header = (): React.JSX.Element => (
  <View style={styles.background}>
    <Text style={styles.headerText}>Mayden Shopping List</Text>
  </View>
);
