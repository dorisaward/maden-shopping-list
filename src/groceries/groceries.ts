import {GroceryItem} from './GroceryItem';

export const groceries: GroceryItem[] = [
  {name: 'broccoli', quantity: 1, id: '0'},
  {name: 'spinach', quantity: 2, id: '1'},
  {name: 'potatoes', quantity: 5, id: '2'},
  {name: 'kidney beans', quantity: 1, id: '3'},
];
