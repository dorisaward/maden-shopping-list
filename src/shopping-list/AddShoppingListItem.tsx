import React, {useCallback, useState} from 'react';
import {GroceryItem} from '../groceries/GroceryItem';
import {View, TextInput, TouchableOpacity, Text, Alert} from 'react-native';
import {styles} from '../styles';
import uuid from 'react-native-uuid';

type Props = {
  addGroceryItem: (groceryItem: GroceryItem) => void;
};

export const AddShoppingListItem = ({
  addGroceryItem,
}: Props): React.JSX.Element => {
  const [newGroceryName, setNewGroceryName] = useState<string | undefined>(
    undefined,
  );
  const [newGroceryQuantity, setNewGroceryQuantity] = useState<number>(0);

  const setQuantity = useCallback(
    (text: string) => {
      try {
        const number = Number(text);
        setNewGroceryQuantity(number);
      } catch (e) {
        console.warn(`Could not cast string ${text} to number`, e);
      }
    },
    [setNewGroceryQuantity],
  );

  const saveGroceryItem = useCallback(() => {
    if (!newGroceryName) {
      Alert.alert('Please make sure your groceries have a name');
      return;
    }
    if (!newGroceryQuantity) {
      Alert.alert(`Please specify how much ${newGroceryName} you want.`);
      return;
    }
    addGroceryItem({
      name: newGroceryName,
      quantity: newGroceryQuantity,
      id: uuid.v4().toString(),
    });
    setNewGroceryQuantity(0);
    setNewGroceryName(undefined);
  }, [newGroceryName, newGroceryQuantity, addGroceryItem]);

  return (
    <View style={styles.shoppingListItem}>
      <TextInput
        placeholder="Name"
        value={newGroceryName}
        style={styles.shoppingListItemInput}
        onChangeText={setNewGroceryName}
      />
      <TextInput
        placeholder="Quantity"
        value={newGroceryQuantity.toString()}
        keyboardType="number-pad"
        style={styles.shoppingListItemInput}
        onChangeText={setQuantity}
      />
      <TouchableOpacity style={styles.saveText} onPress={saveGroceryItem}>
        <Text>Save</Text>
      </TouchableOpacity>
    </View>
  );
};
