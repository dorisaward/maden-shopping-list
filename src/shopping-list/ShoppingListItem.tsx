import React from 'react';
import {Text, TouchableOpacity, View} from 'react-native';
import {GroceryItem} from '../groceries/GroceryItem';
import {styles} from '../styles';

type Props = {
  item: GroceryItem;
  removeGroceryItem: () => void;
};

export const ShoppingListItem = ({item, removeGroceryItem}: Props) => (
  <View style={styles.shoppingListItem}>
    <Text style={styles.shoppingListItemText}>{item.name}</Text>
    <Text style={styles.shoppingListItemText}>{item.quantity}</Text>
    <TouchableOpacity style={styles.removeText} onPress={removeGroceryItem}>
      <Text>Remove</Text>
    </TouchableOpacity>
  </View>
);
