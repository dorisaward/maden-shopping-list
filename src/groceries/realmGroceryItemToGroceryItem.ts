import {GroceryItem} from './GroceryItem';
import {RealmGroceryItem} from './RealmGroceryItem';

export const realmGroceryItemToGroceryItem = (
  realmGroceryItem: RealmGroceryItem,
): GroceryItem => ({
  _id: realmGroceryItem._id,
  name: realmGroceryItem.name,
  quantity: realmGroceryItem.quantity,
});
