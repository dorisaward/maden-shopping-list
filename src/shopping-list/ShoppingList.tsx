import React, {useCallback, useState} from 'react';
import {FlatList, FlatListProps, ListRenderItemInfo} from 'react-native';
import {ShoppingListItem} from './ShoppingListItem';
import {GroceryItem} from '../groceries/GroceryItem';
import {AddShoppingListItem} from './AddShoppingListItem';

type Props = {
  initialGroceryItems: GroceryItem[];
};

export const ShoppingList = ({
  initialGroceryItems,
}: Props): React.JSX.Element => {
  const [groceryItems, setGroceryItems] =
    useState<GroceryItem[]>(initialGroceryItems);
  const addGroceryItem = useCallback(
    (groceryItem: GroceryItem) =>
      setGroceryItems(items => [...items, groceryItem]),
    [setGroceryItems],
  );
  const removeGroceryItem = useCallback(
    (groceryItem: GroceryItem) => () =>
      setGroceryItems(items =>
        items.filter(item => item.id !== groceryItem.id),
      ),
    [setGroceryItems],
  );
  const ListHeaderCooponent = useCallback(
    () => <AddShoppingListItem addGroceryItem={addGroceryItem} />,
    [addGroceryItem],
  );
  const renderItem = useCallback(
    ({item}: ListRenderItemInfo<GroceryItem>) => (
      <ShoppingListItem
        item={item}
        removeGroceryItem={removeGroceryItem(item)}
      />
    ),
    [removeGroceryItem],
  );

  return (
    <FlatList
      data={groceryItems}
      renderItem={renderItem}
      keyExtractor={keyExtractor}
      ListHeaderComponent={ListHeaderCooponent}
    />
  );
};

const keyExtractor: FlatListProps<GroceryItem>['keyExtractor'] = item =>
  item.id;
