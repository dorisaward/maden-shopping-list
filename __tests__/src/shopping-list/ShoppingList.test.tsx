import 'react-native';
import React from 'react';
import {it, expect} from '@jest/globals';
import {render} from '@testing-library/react-native';
import {GroceryItem} from '../../../src/groceries/GroceryItem';
import {ShoppingList} from '../../../src/shopping-list/ShoppingList';

it('should render shopping list', () => {
  // Given
  const item0: GroceryItem = {
    id: 'test-id-0',
    name: 'test name 0',
    quantity: 1,
  };
  const item1: GroceryItem = {
    id: 'test-id-1',
    name: 'test name 1',
    quantity: 2,
  };
  const renderable = <ShoppingList initialGroceryItems={[item0, item1]} />;

  // When
  const {toJSON} = render(renderable);

  // Then
  expect(toJSON()).toMatchSnapshot();
});
