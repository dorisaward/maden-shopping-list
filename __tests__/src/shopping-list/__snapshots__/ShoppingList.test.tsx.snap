// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render shopping list 1`] = `
<RCTScrollView
  ListHeaderComponent={[Function]}
  data={
    [
      {
        "id": "test-id-0",
        "name": "test name 0",
        "quantity": 1,
      },
      {
        "id": "test-id-1",
        "name": "test name 1",
        "quantity": 2,
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      collapsable={false}
      onLayout={[Function]}
    >
      <View
        style={
          {
            "borderColor": "#6495ed",
            "borderRadius": 30,
            "borderWidth": 3,
            "flexDirection": "row",
            "marginBottom": 3,
            "width": "100%",
          }
        }
      >
        <TextInput
          onChangeText={[Function]}
          placeholder="Name"
          style={
            {
              "borderColor": "grey",
              "borderRadius": 30,
              "borderWidth": 1,
              "margin": 5,
              "minWidth": "33%",
              "paddingHorizontal": 5,
              "paddingVertical": 0,
            }
          }
        />
        <TextInput
          keyboardType="number-pad"
          onChangeText={[Function]}
          placeholder="Quantity"
          style={
            {
              "borderColor": "grey",
              "borderRadius": 30,
              "borderWidth": 1,
              "margin": 5,
              "minWidth": "33%",
              "paddingHorizontal": 5,
              "paddingVertical": 0,
            }
          }
          value="0"
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "justifyContent": "center",
              "opacity": 1,
              "paddingLeft": 10,
              "textAlign": "right",
            }
          }
        >
          <Text>
            Save
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "borderColor": "#6495ed",
            "borderRadius": 30,
            "borderWidth": 3,
            "flexDirection": "row",
            "marginBottom": 3,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 15,
              "minWidth": "33%",
              "padding": 10,
            }
          }
        >
          test name 0
        </Text>
        <Text
          style={
            {
              "fontSize": 15,
              "minWidth": "33%",
              "padding": 10,
            }
          }
        >
          1
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "flexShrink": 1,
              "justifyContent": "center",
              "opacity": 1,
              "paddingLeft": 10,
              "textAlign": "right",
            }
          }
        >
          <Text>
            Remove
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "borderColor": "#6495ed",
            "borderRadius": 30,
            "borderWidth": 3,
            "flexDirection": "row",
            "marginBottom": 3,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 15,
              "minWidth": "33%",
              "padding": 10,
            }
          }
        >
          test name 1
        </Text>
        <Text
          style={
            {
              "fontSize": 15,
              "minWidth": "33%",
              "padding": 10,
            }
          }
        >
          2
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "flexShrink": 1,
              "justifyContent": "center",
              "opacity": 1,
              "paddingLeft": 10,
              "textAlign": "right",
            }
          }
        >
          <Text>
            Remove
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
